rfcs/0000-plan-dynanism.md

```diff
@@ -0,0 +1,251 @@
+Many languages now support very large library ecosystems, with dependencies expressed in a language-specific package manager.
+To this new generation of developers, the distro (or homebrew) is a crufty relic from an earlier age to bootstrap modernity, and then be forgotten about.
+
+Right now, to deal with these packages, we either convert by hand, or commit lots of generated code into Nixpkgs.
+But I don't think either of those options are healthy or sustainable.
+The problem with the first is sheer effort; we'll never be able to keep up.
+The problem with the second is bloating Nixpkgs but more importantly reproducability: If someone wants to update that generated code it is unclear how.
+All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.
+
+The core feature here, derivations that build derivations, is the best fundamental primitive for this problem.
+It's very performant, being well-adapted for Nix's current scheduler.
+Unlike recursive Nix, there's is no potential for half-built dependencies to sit around waiting for other builds, wasting resources.
+Each build step (derivation) always runs start to finish blocking on nothing.
+It's very efficient, because it doesn't obligate the use of the Nix expression language.
+Finally, it's quite compatible with `--dry-run`.
+
+However, "import from derivation" is still far and away the easiest method to use, and the one that existing tools to convert to Nix use.
+We should continue to support it, finding a way for `hydra.nixos.org` to allow it, so those tools with IFD can be used in Nixpkgs and become first-class members of the Nix ecosystem.
+We have a fairly straightforward mechanism, only slightly more cumbersome than IFD today, to allow evaluation to be deferred after imported things are built.
+This frees up the Hydra evaluator to finish before building, and also meshes well with any plan to build more than one eval-realized path at a time.
+This should allow us to drop the `hydra.nixos.org` restriction.
+
+With these steps, I think we will be able to successfully convert to Nix a bunch of developers that mainly work in one language, and didn't even think they were in need of a better distro.
+In turn, I hope these upstream packages and ecosystems might even care about packaging and integration of the sort that we do.
+This would create a virtuous cycle where Nix is easier to use by more people, and Nixpkgs is easier to maintain as upstream packages better match our values.
+
+# Detailed design
+[design]: #detailed-design
+
+We can break this down nicely into steps.
+
+## Dynamic derivations
+
+*This is implemented in https://github.com/NixOS/nix/pull/4628.*
+
+1. Derivation outputs can be valid derivations:
+
+   1. Allow derivation outputs to be content addressed with the "text hashing" scheme.
+
+   2. Lift the restriction barring derivations output paths from ending in `.drv` if they are so content-addressed
```

Comment:
I'm also not quite understanding this. Are you saying that currently a Nix derivation can't have an output that ends in `.drv`? Again, is this documented anywhere?

Reply:
Doesn't seem to be documented in the nix manual. This is the check: <https://github.com/NixOS/nix/blob/57f321f3ce311228ad9ab7eac8b108aeb4fa4026/src/libexpr/primops.cc#L1110-L1115>

Reply:
It might just help with the reading to put in a little additional glue in this doc regarding CA.

I guess the key idea, which is not explicitly mentioned is:

> .drv can be content addressed by hashing their canonical (?) text representation and since those text instructions themselves produce reproducible outputs can be subsituted during evaluation in lieu of the actual outputs.

Reply:
> themselves produce reproducible outputs can be subsituted during evaluation in lieu of the actual outputs

This is however still a leap of faith due to input-addressing (as in the extensional model which is described [in chapter 5 here](https://edolstra.github.io/pubs/phd-thesis.pdf)).

Reply:
Tried to expand on the big picture with https://github.com/NixOS/rfcs/pull/92/files#r646174922, does that solve it?

Reply:
That link is dead by now, maybe can you link to a commit diff?

Reply:
This is fixed now.